<h1> To remove duplicate images from a given dataset </h1>
<ol>
    <li> Average Hashing    (aHash) </li>
    <li> Peceptual Hashing (pHash) </li>
    <li> Difference Hashing (dHash) </li>
</ol>

In [1]:
import cv2
import os
import numpy as np
import matplotlib.pyplot as plt
from scipy.fftpack import dct
import csv

# using aHash

1. Import Images
2. Reduce there sizes to 8x8
3. convert image to grayscale
4. Average the colors
5. Based on whether an individual pixel more than average value , we assign it a 1/0 value

In [2]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(8,8))
    img.append(dogImgResized)

In [3]:
#Computing hash values for each img
aHash  =[]
for dog in img:
    mean_color = np.mean(dog)
    new_mat = [x>mean_color for x in dog]
    new_mat = np.array(new_mat,dtype = 'int').reshape((-1,))
    binVal = ''.join(str(x) for x in new_mat)
    hashVal = hex(int(binVal, 2)) # for reference
    aHash.append(binVal)

In [4]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = aHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [5]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        # Printing some duplicate names
        if(duplicated_names <= 20 ):
            print(duplicated[keys])
# # Saving the result
with open('aHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg', 'n@dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg']
['c_dog.1000.jpg', 'dog.1000.jpg', 'n@dog.1000.jpg']
['c_dog.1001.jpg', 'dog.1001.jpg']
['c_dog.1002.jpg', 'dog.1002.jpg', 'n@dog.1002.jpg']
['c_dog.1003.jpg', 'dog.1003.jpg']
['c_dog.1004.jpg', 'dog.1004.jpg', 'n@dog.1004.jpg']
['c_dog.1005.jpg', 'dog.1005.jpg', 'n@dog.1005.jpg']
['c_dog.1006.jpg', 'dog.1006.jpg']
['c_dog.1007.jpg', 'dog.1007.jpg']
['c_dog.1008.jpg', 'dog.1008.jpg', 'n@dog.1008.jpg']
['c_dog.1009.jpg', 'dog.1009.jpg', 'n@dog.1009.jpg']
['c_dog.101.jpg', 'dog.101.jpg']
['c_dog.1010.jpg', 'dog.1010.jpg']
['c_dog.1011.jpg', 'dog.1011.jpg']
['c_dog.1012.jpg', 'dog.1012.jpg']
['c_dog.1013.jpg', 'dog.1013.jpg']
['c_dog.1014.jpg', 'dog.1014.jpg']
['c_dog.1015.jpg', 'dog.1015.jpg']
['c_dog.1016.jpg', 'dog.1016.jpg', 'n@dog.1016.jpg']


In [6]:
print(duplicated_items,duplicated_names,total_images)

2383 1001 6432


<p> Our analysis using aHash reveals that there are a total of 1001 uniques images(out of 6432 unique images) which have been duplicated a grand total of 2383 </p>

In [7]:
# Visualizing duplicates
itr = 0
for keys in duplicated.keys():
    if(itr == 10 ):
        break
    if(len(duplicated[keys]) >= 2):
        itr+=1;
        print("Encountered Duplicates")
        for values in duplicated[keys]:
            filePath = os.path.join('aug_dogs/',values)
            curr_img = cv2.imread(filePath)
            cv2.imshow(str(itr),curr_img)
            cv2.waitKey(0)
            cv2.destroyAllWindows()

Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates
Encountered Duplicates


# using pHash

1. Import Images
2. Reduce there sizes to 32x32
3. convert image to grayscale
4. Convert image to its dct eqv
5. Crop the upper 8x8 half
6. Take average of pixel values
7. Iterate through all the pixels and if they are pixel values > average then make them 1 else 0

In [8]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(32,32))
    img.append(dogImgResized)

In [9]:
#Computing hash values for each img
pHash  =[]
for dog in img:
    
    dctImage = dct(dog,type = 2)[:8,:8]
    upperHalfDct = (dctImage.sum() - dctImage[0][0])/63.0
    dctImage = dctImage.reshape((-1,))
    dctImage = [x>upperHalfDct for x in dctImage]
    dctImage = np.array(dctImage,dtype = 'int')
    dctImage = dctImage.reshape((-1,))
    binVal = ''.join(str(x) for x in dctImage)
    hashVal = hex(int(binVal, 2)) # for reference
    pHash.append(binVal)

In [10]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = pHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [11]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        # Printing some duplicate names
        if(duplicated_names <= 20 ):
            print(duplicated[keys])
# # Saving the result
with open('pHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg']
['c_dog.1000.jpg', 'c_dog.3137.jpg', 'dog.1000.jpg', 'f0@dog.1816.jpg', 'f1@dog.1408.jpg', 'n@dog.1000.jpg']
['c_dog.1001.jpg', 'c_dog.1002.jpg', 'c_dog.1004.jpg', 'c_dog.1014.jpg', 'c_dog.1018.jpg', 'c_dog.1023.jpg', 'c_dog.1046.jpg', 'c_dog.1048.jpg', 'c_dog.105.jpg', 'c_dog.1057.jpg', 'c_dog.1058.jpg', 'c_dog.1064.jpg', 'c_dog.1065.jpg', 'c_dog.107.jpg', 'c_dog.1076.jpg', 'c_dog.1079.jpg', 'c_dog.1083.jpg', 'c_dog.1090.jpg', 'c_dog.1091.jpg', 'c_dog.1095.jpg', 'c_dog.11.jpg', 'c_dog.1100.jpg', 'c_dog.1112.jpg', 'c_dog.1113.jpg', 'c_dog.1114.jpg', 'c_dog.1117.jpg', 'c_dog.113.jpg', 'c_dog.1132.jpg', 'c_dog.1140.jpg', 'c_dog.1146.jpg', 'c_dog.1156.jpg', 'c_dog.1162.jpg', 'c_dog.1169.jpg', 'c_dog.1173.jpg', 'c_dog.1195.jpg', 'c_dog.1206.jpg', 'c_dog.1208.jpg', 'c_dog.1210.jpg', 'c_dog.1214.jpg', 'c_dog.1216.jpg', 'c_dog.1221.jpg', 'c_dog.1226.jpg', 'c_dog.1242.jpg', 'c_dog.1243.jpg', 'c_dog.1257.jpg', 'c_dog.1270.jpg', 'c_

In [12]:
x =['c_dog.1000.jpg', 'c_dog.3137.jpg', 'dog.1000.jpg', 'f0@dog.1816.jpg', 'f1@dog.1408.jpg', 'n@dog.1000.jpg']
for item in x:
    curr_img = cv2.imread('aug_dogs/' + item)
    cv2.imshow('img',curr_img)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

In [13]:
print(total_images,duplicated_names,duplicated_items)

3746 853 4921


Acc to pHash , there are a total of 3746 unique images with 853 duplicated images over 4921 times
<strong> 
    This Does not look correct as I have visualized the results and they dont look correct( However there is a change that i could have analysed the misclassifications )
</strong>
This Could be corrected if we use similiarity between hamming codes , but that would help us in finding whether two images are similar or not , it will not help us in grouping all the images which are duplicates

or it could be that pHash tries to find generalization in images

# using dHash

1. Import Images
2. Reduce there sizes to 9x8
3. convert image to grayscale
4. Averaging the color
5. from index 1 .. 8 compare each cell in each row , with its previous cell in the same row , assign bit values on the relative magnitude of these 2 values

In [14]:
# Importing images

img = []   # Names of the images
names = [] # Names of image files
ctr =0
for dog in os.listdir('aug_dogs'):
    names.append(dog)
    ctr+=1
    filePath = os.path.join('aug_dogs/',dog)
    dogImg = cv2.imread(filePath)
    #Converting image to gray scale
    dogImgGrayScale = cv2.cvtColor(dogImg,cv2.COLOR_BGR2GRAY)
    #Resizing image
    dogImgResized = cv2.resize(dogImgGrayScale,(9,8))
    img.append(dogImgResized)

In [15]:
#Computing hash values for each img
dHash  =[]
for dog in img:
    
    x1 = dog[:,0:8]
    x2 = dog[:,1:9]

    gradient = np.array((x1>x2),dtype = 'int')    
    gradient = gradient.reshape((-1,))
    binVal = ''.join(str(x) for x in gradient)
    hashVal = hex(int(binVal, 2)) # for reference
    dHash.append(binVal)


In [16]:
# Making a dictionary of values having same bits
duplicated = {}
for i in range(len(names)):
    curr_name = names[i]
    curr_hash = dHash[i]
    try:
        duplicated[curr_hash].append(curr_name)
    except:
        duplicated[curr_hash] = []
        duplicated[curr_hash].append(curr_name)

In [17]:
# Analyzing the results
duplicated_names = 0
duplicated_items = 0
total_images =0 
for keys in duplicated.keys():
    total_images+=1
    if(len(duplicated[keys]) >= 2):
        duplicated_names+=1
        duplicated_items += len(duplicated[keys])
        # Printing some duplicate names
        if(duplicated_names <= 20 ):
            print(duplicated[keys])
        
# # Saving the result
with open('dHash.csv', 'w') as f:
    for key in duplicated.keys():
        for elem in duplicated[key]:
            f.write("%s,"%(elem))
        f.write("\n")

['c_dog.10.jpg', 'dog.10.jpg']
['c_dog.100.jpg', 'dog.100.jpg', 'n@dog.100.jpg']
['c_dog.1000.jpg', 'dog.1000.jpg']
['c_dog.1001.jpg', 'dog.1001.jpg']
['c_dog.1002.jpg', 'dog.1002.jpg']
['c_dog.1003.jpg', 'dog.1003.jpg']
['c_dog.1004.jpg', 'dog.1004.jpg']
['c_dog.1005.jpg', 'dog.1005.jpg']
['c_dog.1006.jpg', 'dog.1006.jpg']
['c_dog.1007.jpg', 'dog.1007.jpg']
['c_dog.1008.jpg', 'dog.1008.jpg']
['c_dog.1009.jpg', 'dog.1009.jpg']
['c_dog.101.jpg', 'dog.101.jpg']
['c_dog.1010.jpg', 'dog.1010.jpg']
['c_dog.1011.jpg', 'dog.1011.jpg']
['c_dog.1012.jpg', 'dog.1012.jpg', 'n@dog.1012.jpg']
['c_dog.1013.jpg', 'dog.1013.jpg']
['c_dog.1014.jpg', 'dog.1014.jpg']
['c_dog.1015.jpg', 'dog.1015.jpg']
['c_dog.1016.jpg', 'dog.1016.jpg']


In [18]:
print(total_images)
print(duplicated_names)
print(duplicated_items)

6700
1001
2115


A total of 1001 images out of 6700 have been duplicated a total of 2115 times

<h3> Conclusions </h3>
<hr>
1.aHash -  Our analysis using aHash reveals that there are a total of 1001 uniques images(out of 6432 unique images) which have been duplicated a grand total of 2383 

2.pHash - Acc to pHash , there are a total of 3746 unique images with 853 duplicated images over 4921 times This Does not look correct as I have visualized the results and they dont look correct( However there is a change that i could have analysed the misclassifications ) This Could be corrected if we use similiarity between hamming codes , but that would help us in finding whether two images are similar or not , it will not help us in grouping all the images which are duplicates

3.dHash - A total of 1001 images out of 6700 have been duplicated a total of 2115 times